text/0000-buildpack-uris.md

```diff
@@ -0,0 +1,172 @@
+| CNB Builder | `cnb:builder://[<host>]/[<id>[:<version>]]` | `cnb:builder://`<br>`cnb:builder:///bp.id`<br>`cnb:builder:///bp.id:bp.version`
+
+
+### Relative
+
+A URI without a scheme is treated as a relative reference as per [RFC 3986 section 4.1](https://tools.ietf.org/html/rfc3986#section-4.1).
+
+### Filesystem
+
+The filesystem scheme is the implementation of [RFC 8089](https://tools.ietf.org/html/rfc8089).
+
+### URL
+
+The URL scheme is the implementation of [RFC 7230](https://tools.ietf.org/html/rfc7230#section-2.7.1).
+
+### Docker
+
+The docker scheme denotes the use of the [Docker HTTP API v2 protocol](https://docs.docker.com/registry/spec/api/).
+
+- `<host>` is optional and defaults to `index.docker.io`.
+- Similar to the `file` scheme, there is a minimal declaration for omitting host by using a simple slash (`/`). 
+eg. `docker:/ubuntu:latest`
+
+### CNB Builder
+
+- `cnb:builder://` - A reference to ALL the buildpacks in the builder.
+- `cnb:builder:///bp.id` - A reference to a buildpack with id `bp.id` in the builder. If there is more than one version
+available, a error should be thrown.
+- `cnb:builder:///bp.id:bp.version` - A reference to a buildpack with id `bp.id` at the specific version `bp.version`.
+- Similar to the `file` scheme, there is a minimal declaration for omitting host by using a simple slash (`/`). 
+eg. `cnb:builder:/bp.id:bp.version`
+
+# How it Works
+[how-it-works]: #how-it-works
+
+### pack
+
+Certain `pack` CLI arguments would take URIs supported instead of the current inconsistent format.
+
+Some examples:
+
+```shell script
+pack build my-app \
+  --buildpack cnb:builder://bp.id@bp.version \                # buildpack from builder
+  --buildpack file:///home/user/path/to/buildpack/ \          # absolute via file
+  --buildpack /home/user/path/to/buildpack/ \                 # absolute via schemeless
+  --buildpack ../path/to/buildpack/ \                         # relative file
+  --buildpack docker:/cnbs/some-package \                     # Docker Hub image
+  --buildpack docker://gcr.io/cnbs/sample-package-2:bionic \  # GCR image (with tag)
+  --buildpack cnb:builder://                                  # All buildpacks in builder
+```   
+
+### Config Files
+
+Config files would replace `image` with `uri`, using the `docker://` scheme, where applicable.
+
+`package.toml`:
+```toml
+[buildpack]
+uri = "."
+
+[[dependencies]]
+uri = "docker:/my/image:latest"
+
+# ...
+``` 
+
+`builder.toml`:
+```toml
+# ...
+
+[[buildpacks]]
+uri = "./my/buildpacks/"
+
+[[buildpacks]]
+uri = "docker:/my/image:latest"
+
+# ...
+```
+
+# Additional considerations
+
+#### `pack` User Expirience
+
+`pack` could provide additional logic in order for users to have a nicer user experience when providing references
+via CLI. For example, `pack` could have an order of precedence to make a best guess at which type of reference
+they are referring to or provide additional feedback if it's ambiguous.
```

Comment:
I think I'd recommend this order:

1. builder image
2. registry
3. file system path